flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/stream/StreamingSemiAntiJoinOperator.java

```diff
@@ -0,0 +1,196 @@
+	 *   if semi join, send input record
+	 * if the input record is accumulate, state.add(record, matched size)
+	 * if the input record is retract, state.retract(record)
+	 * </pre>
+	 */
+	@Override
+	public void processElement1(StreamRecord<BaseRow> element) throws Exception {
+		BaseRow input = element.getValue();
+		AssociatedRecords associatedRecords = AssociatedRecords.of(input, true, rightRecordStateView, joinCondition);
+		if (associatedRecords.isEmpty()) {
+			if (isAntiJoin) {
+				collector.collect(input);
+			}
+		} else { // there are matched rows on the other side
+			if (!isAntiJoin) {
+				collector.collect(input);
+			}
+		}
+		if (BaseRowUtil.isAccumulateMsg(input)) {
+			leftRecordStateView.addRecord(input, associatedRecords.size());
+		} else { // input is retract
+			input.setHeader(BaseRowUtil.ACCUMULATE_MSG);
+			leftRecordStateView.retractRecord(input);
+		}
+	}
+
+	/**
+	 * Process an input element and output incremental joined records, retraction messages will
+	 * be sent in some scenarios.
+	 *
+	 * <p>Following is the pseudo code to describe the core logic of this method.
+	 *
+	 * <pre>
+	 * if input record is accumulate
+	 * | state.add(record)
+	 * | if there is no matched rows on the other side, skip
+	 * | if there are matched rows on the other side
+	 * | | if the matched num in the matched rows == 0
+	 * | |   if anti join, send -[other]s
+	 * | |   if semi join, send +[other]s
+	 * | | if the matched num in the matched rows > 0, skip
+	 * | | otherState.update(other, old+1)
+	 * | endif
+	 * endif
+	 * if input record is retract
+	 * | state.retract(record)
+	 * | if there is no matched rows on the other side, skip
+	 * | if there are matched rows on the other side
+	 * | | if the matched num in the matched rows == 0, this should never happen!
+	 * | | if the matched num in the matched rows == 1
+	 * | |   if semi join, send -[other]
+	 * | |   if anti join, send +[other]
+	 * | | if the matched num in the matched rows > 1, skip
+	 * | | otherState.update(other, old-1)
+	 * | endif
+	 * endif
+	 * </pre>
+	 */
+	@Override
+	public void processElement2(StreamRecord<BaseRow> element) throws Exception {
+		BaseRow input = element.getValue();
+		AssociatedRecords associatedRecords = AssociatedRecords.of(input, false, leftRecordStateView, joinCondition);
+		if (BaseRowUtil.isAccumulateMsg(input)) {
+			rightRecordStateView.addRecord(input);
+			if (!associatedRecords.isEmpty()) {
+				// there are matched rows on the other side
+				for (OuterRecord outerRecord : associatedRecords.getOuterRecords()) {
+					BaseRow other = outerRecord.record;
+					if (outerRecord.numOfAssociations == 0) {
+						if (isAntiJoin) {
+							// send -[other]
+							other.setHeader(BaseRowUtil.RETRACT_MSG);
+						} else {
+							// send +[other]
+							other.setHeader(BaseRowUtil.ACCUMULATE_MSG);
+						}
+						collector.collect(other);
+					} // ignore when number > 0
+					other.setHeader(BaseRowUtil.ACCUMULATE_MSG);
```

Comment:
why set acc to other? Will it be used again?

Reply:
Header of `other` may be set to `RETRACT` before, so we have to set it back when storing it in state. All rows stored in state are all accumulate BaseRows.

I did a minor adjust for this to move this line close to setting retract header, and add a comment, to make it more readable.